libslide/src/partial_evaluator/flatten.rs

```diff
@@ -0,0 +1,139 @@
+
+pub fn flatten_expr(expr: &Rc<Expr>) -> Rc<Expr> {
+    match expr.as_ref() {
+        Expr::Const(_) | Expr::Var(_) => Rc::clone(expr),
+        Expr::Parend(inner) | Expr::Bracketed(inner) => Rc::clone(inner),
+        Expr::BinaryExpr(BinaryExpr { op, lhs, rhs })
+            if op == &BinaryOperator::Plus || op == &BinaryOperator::Minus =>
+        {
+            flatten_add_or_sub(lhs, rhs, op == &BinaryOperator::Minus)
+        }
+        // TODO: handle everything else
+        _ => Rc::clone(expr),
+    }
+}
+
+pub fn flatten_add_or_sub(o_lhs: &Rc<Expr>, o_rhs: &Rc<Expr>, is_subtract: bool) -> Rc<Expr> {
+    let lhs = flatten_expr(o_lhs);
+    let rhs = flatten_expr(o_rhs);
+
+    let mut coeff = 0f64;
+    let mut terms = HashMap::<&Rc<Expr>, f64>::new();
+    let mut args = VecDeque::with_capacity(2);
+    let base_args = [lhs, rhs];
+    args.extend(base_args.iter());
+    let mut args_before_neg = 1;
+    while let Some(arg) = args.pop_front() {
+        let is_neg = is_subtract && args_before_neg == 0;
+        args_before_neg -= 1;
+
+        match arg.as_ref() {
+            Expr::Const(konst) => {
+                if is_neg {
+                    coeff -= konst;
+                } else {
+                    coeff += konst;
+                }
+            }
+            Expr::BinaryExpr(BinaryExpr {
+                op: BinaryOperator::Plus,
+                lhs,
+                rhs,
+            }) => {
+                if is_neg {
+                    args.push_back(lhs);
+                    args.push_back(rhs);
+                } else {
+                    args.push_front(lhs);
+                    args.push_front(rhs);
+                    args_before_neg += 2;
+                }
+            }
+            Expr::Var(_) | _ => {
+                // TODO: handle everything else more granularly
+                let entry = terms.entry(arg).or_insert(0.);
+                if is_neg {
+                    *entry -= 1.;
+                } else {
+                    *entry += 1.;
+                }
+            }
+        }
+    }
+
+    let mut new_args: Vec<Rc<Expr>> = Vec::with_capacity(1 + terms.len());
+    if coeff != 0. {
+        new_args.push(Rc::from(Expr::Const(coeff)));
+    }
+    for (term, coeff) in terms.into_iter() {
+        if coeff == 0. {
+            continue;
+        } else if coeff == 1. {
+            new_args.push(Rc::clone(term));
+        } else if coeff == -1. {
+            let neg = UnaryExpr::negate(Rc::clone(term));
+            new_args.push(Rc::from(Expr::UnaryExpr(neg)));
+        } else {
+            let mult = BinaryExpr::mult(Expr::Const(coeff), Rc::clone(term));
+            let expr: Expr = mult.into();
+            new_args.push(Rc::from(expr));
+        }
+    }
+
+    match new_args.len() {
+        0 => Rc::from(Expr::Const(0.)),
+        1 => new_args.into_iter().next().unwrap(),
+        _ => unflatten_binary_expr(&new_args, BinaryOperator::Plus, UnflattenStrategy::Left),
```

Comment:
why don't you just use get instead of casting to an iter and unwrapping in new_args

Reply:
`get` returns a reference, but we don't want that, we want to return the value, so it's find to destroy the vec

Reply:
Getting a reference and casting it is faster then casting iter element and then getting a next and then unwrapping though

Reply:
Getting a reference and casting it is faster then casting iter element and then getting a next and then unwrapping though

Reply:
wait what? casting how? We need to go from a reference to a raw value

Reply:
I thought you can just dereference

Reply:
No cuz that would move the value out of the vector which is unsound.

Reply:
Why is it unsound if your destroying the deque anyway

Reply:
yeah but the compiler isn't smart enough to know that: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=47cedd7a1acf53f5bff12a0461a41f74

Reply:
There is a better way to do this though, you're right: remove the value from the vector with `swap_remove`